packages/world/src/Constants.sol

```diff
@@ -94,3 +87,18 @@ uint256 constant MAX_NEPTUNIUM = 18_696;
 
 uint128 constant INITIAL_ENERGY_PER_VEGETATION = 10000000000000000;
 uint128 constant INITIAL_LOCAL_ENERGY_BUFFER = 131480700000000000000;
+
+// Rate limit constants
+uint128 constant MAX_RATE_LIMIT_UNITS_PER_BLOCK = 1e18;
+uint128 constant MAX_RATE_LIMIT_UNITS_PER_SECOND = MAX_RATE_LIMIT_UNITS_PER_BLOCK / BLOCK_TIME;
+
+// Movement rate limits
+uint128 constant WALK_UNIT_COST = MAX_RATE_LIMIT_UNITS_PER_SECOND / 15; // 15 blocks per second
+uint128 constant SWIM_UNIT_COST = MAX_RATE_LIMIT_UNITS_PER_SECOND * 10 / 135; // 13.5 blocks per second (90% of walking speed)
+
+// Combat rate limits
+uint128 constant HIT_UNIT_COST = MAX_RATE_LIMIT_UNITS_PER_BLOCK; // 1 hit per block
```

Comment:
hmm how will 1 hit per block with preconfs?

Reply:
it should just revert, we should probably handle the limit in the client